### PR TITLE
Rework clustering

### PR DESCRIPTION
**Pull request by FabioLuporini**
A new clustering algorithm, which generalises what we have in master.

- Support more corner use cases
- Smarter equation grouping (better "loop fusion")
- Easier to understand
- More robust
- Easier to extend

This PR also ships a couple of fixes related to the DSE, which however rely on the new clustering algorithm

### CODE REVIEW

File: devito/dle/rewriters.py [hunk 1]
```diff
@@ -18,6 +17,7 @@
 from devito.logger import perf_adv
 from devito.mpi import HaloExchangeBuilder
 from devito.parameters import configuration
+from devito.symbolics import ccode
```
Comment (georgebisbas) [inline devito/dle/rewriters.py]: ?

File: devito/dse/rewriters.py [hunk 2]
```diff
@@ -320,35 +321,28 @@ def _eliminate_inter_stencil_redundancies(self, cluster, template, **kwargs):
...
+            # induce a flow/anti dependence (below, `i.limits` captures this, by
+            # telling how much halo will be needed to honour such dependences)
+            dep_inducing = [i for i in writeto if any(i.limits)]
+            try:
+                index = writeto.index(dep_inducing[0])
+                writeto = IntervalGroup(writeto[index:])
+            except IndexError:
+                warning("Couldn't optimize some of the detected redundancies")
 
             # Create a temporary to store `alias`
```
Comment (georgebisbas) [inline devito/dse/rewriters.py]: temporary 'xxxx' ?

File: devito/dse/transformer.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
-from devito.ir.clusters import ClusterGroup, groupby
+from devito.ir.clusters import ClusterGroup, optimize
```
Comment (georgebisbas) [inline devito/dse/transformer.py]: Is the 'optimize' a bit vague?

File: devito/ir/clusters/algorithms.py [hunk 4]
```diff
@@ -1,254 +1,369 @@
...
+        # Take the innermost Dimension -- no other Clusters other than those in
+        # `clusters` are supposed to share it
+        candidates = prefix[-1].dim._defines
+
+        scope = Scope(exprs=flatten(c.exprs for c in clusters))
+
+        # The most nasty case:
+        # eq0 := u[t+1, x] = ... u[t, x]
+        # eq1 := v[t+1, x] = ... v[t, x] ... u[t, x] ... u[t+1, x] ... u[t+2, x]
+        # Here, `eq0` marches forward along `t`, while `eq1` has both a flow and an
```
Comment (georgebisbas) [inline devito/ir/clusters/algorithms.py]: flow- and anti-

File: devito/ir/clusters/algorithms.py [hunk 5]
```diff
@@ -1,254 +1,369 @@
...
+        # anti dependence with `eq0`, which ultimately will require `eq1` to go in
+        # a separate t-loop
+        require_flow_break = (scope.d_flow.cause & scope.d_anti.cause) & candidates
+        if require_flow_break and len(clusters) > 1:
+            backlog = [clusters[-1]] + (backlog or [])
+            # Try with increasingly smaller Cluster groups until the ambiguity is solved
```
Comment (georgebisbas) [inline devito/ir/clusters/algorithms.py]: increasingly smaller?
Maybe gradually ?

File: devito/ir/clusters/algorithms.py [hunk 6]
```diff
@@ -1,254 +1,369 @@
...
+        processed = []
+        for c in clusters:
+            ispace = IterationSpace(c.ispace.intervals, c.ispace.sub_iterators,
+                                    {**c.ispace.directions, **direction})
+            processed.append(Cluster(c.exprs, ispace, c.dspace))
+
+        if backlog is None:
+            return processed
+
+        # Handle the backlog -- the Clusters characterized by flow+anti dependences along
```
Comment (georgebisbas) [inline devito/ir/clusters/algorithms.py]: flow- and anti-

File: devito/ir/clusters/cluster.py [hunk 7]
```diff
@@ -50,17 +69,9 @@ def exprs(self):
...
-        return self._ispace.dimensions
-
     @property
     def itintervals(self):
-        return self._ispace.itintervals
-
-    @property
-    def size(self):
-        return self.ispace.size
+        return self.ispace.itintervals
```
Comment (georgebisbas) [inline devito/ir/clusters/cluster.py]: ? Why is that?

File: devito/ir/clusters/cluster.py [hunk 8]
```diff
@@ -116,12 +137,12 @@ def dtype(self):
...
     def ops(self):
         """The Cluster operation count."""
-        return self.size*sum(estimate_cost(i) for i in self.exprs)
+        return self.ispace.size*sum(estimate_cost(i) for i in self.exprs)
 
-    @property
+    @cached_property
     def traffic(self):
         """
         The Cluster compulsary traffic (number of reads/writes), as a mapper
```
Comment (georgebisbas) [inline devito/ir/clusters/cluster.py]: compulsory

File: devito/ir/support/basic.py [hunk 9]
```diff
@@ -627,6 +629,14 @@ def __init__(self, source, sink):
         self.source = source
         self.sink = sink
 
+    def __eq__(self, other):
+        # If the timestamps are equal in `self` (ie, an inplace dependence) then
```
Comment (georgebisbas) [inline devito/ir/support/basic.py]: i.e.

File: devito/ir/support/basic.py [hunk 10]
```diff
@@ -944,11 +971,11 @@ def d_output(self):
 
     @cached_property
     def d_all(self):
-        """Retrieve all flow, anti, and output dependences."""
+        """Generate all flow, anti, and output dependences."""
```
Comment (georgebisbas) [inline devito/ir/support/basic.py]: dashes?

File: devito/dle/rewriters.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/dle/rewriters.py]: I have just reshuffled things around

File: devito/dse/transformer.py (hunk 3 above)
Comment (FabioLuporini) [reply devito/dse/transformer.py]: maybe yes, but it's pulled from ``clusters`` so it should be clear that it's "optimizing" clusters

File: devito/ir/clusters/cluster.py (hunk 7 above)
Comment (FabioLuporini) [reply devito/ir/clusters/cluster.py]: what?

File: devito/ir/clusters/cluster.py (hunk 8 above)
Comment (FabioLuporini) [reply devito/ir/clusters/cluster.py]: :+1:

File: devito/ir/clusters/cluster.py [hunk 11]
```diff
@@ -148,116 +171,48 @@ def traffic(self):
...
 
-    @exprs.setter
-    def exprs(self, val):
-        self._exprs = val
 
-    @ispace.setter
-    def ispace(self, val):
-        self._ispace = val
-
-    @dspace.setter
```
Comment (FabioLuporini) [inline devito/ir/clusters/cluster.py]: @georgebisbas as promised, these are really going away!

**Comment by FabioLuporini** [comment]
Merged.

**Pull request merged**